### PR TITLE
#42 Discover and load dependencies via composer only after having discovered all source symbols

### LINKED ISSUE #42: Comparison is not just restricted to the source directory
As per recent finding, the comparisons are not currently limited to `src/` (or whatever source path is given), but all classes are being checked.

This means that diffing a codebase will likely cause diffing of all detected classes anywhere if we ever include autoloading of other symbols.

A simple applicable restriction is to only consider classes that have a file path in the defined source directory.

### LINKED ISSUE #47: Move `roave/security-advisories` to `require-dev`
This package can both be used as an application and as a dependency, so adding `roave/security-advisories` to `require` is quite a heavy transitive dependency.

Should be moved to `require-dev` so that it is only active when the package is used as an application.

### PR DESCRIPTION
**Pull request by Ocramius**
Fixes #42
Fixes #47 (done it since I was touching `composer.json` anyway)

### CODE REVIEW

File: src/Comparator.php [hunk 1]
```diff
@@ -23,11 +24,20 @@ public function __construct(ClassBased $classBasedComparisons)
         $this->classBasedComparisons = $classBasedComparisons;
     }
 
-    public function compare(ClassReflector $oldApi, ClassReflector $newApi) : Changes
-    {
+    public function compare(
+        ClassReflector $definedApi,
+        ClassReflector $oldApi,
+        ClassReflector $newApi
```
[the remark this answers is not in the document]
Comment (Ocramius) [reply src/Comparator.php]: No, we need it with dependencies.

`$definedApi` => old classes, no dependencies - only defined symbols
`$oldApi` => old classes with dependencies - needed to analyse inheritance, covariance and such
`$new` => new classes with dependencies - needed to analyse inheritance, covariance and such

To make it more simple: you can't do any operations on the `ReflectionClass` instances without having all the dependencies. Even a simple `ReflectionClass#getParentClasses()` call can lead to a crash.

File: src/Comparator.php [hunk 2]
```diff
@@ -23,11 +24,20 @@ public function __construct(ClassBased $classBasedComparisons)
         $this->classBasedComparisons = $classBasedComparisons;
     }
 
-    public function compare(ClassReflector $oldApi, ClassReflector $newApi) : Changes
-    {
+    public function compare(
+        ClassReflector $definedApi,
```
[the remark this answers is not in the document]
Comment (Ocramius) [reply src/Comparator.php]: I will likely rename the variables to something obnoxiously clear :-)

**Comment by asgrim** [comment]
Also needs a rebase :tumbler_glass:

**Comment by Ocramius** [comment]
@asgrim rebased/fixed/etc - if you merge this I'll try focusing on bells and whistles this evening, to make this nice and shiny for :ship:

**Pull request merged**